src/backend/varstats.c

```diff
@@ -84,127 +45,286 @@ bool VarStatistics::isLexicalScopeVar(symbol* sa)
     if (sa->lnoscopeend > funcsym_p->Sfunc->Fendline.Slinnum)
         return false;
 
-    // verify that the variable does not violate the assumption that scopes are stacked inside
-    // each other (might happen with optimizations/inlining)
-    int start = (sa->lnoscopestart < firstVarStatsLine ? firstVarStatsLine : sa->lnoscopestart + 1);
-    int end   = firstVarStatsLine + cntUsedVarStats;
-    end = (sa->lnoscopeend < end ? sa->lnoscopeend : end);
-    for (int i = start; i < end; i++)
-        if (varStats[i - firstVarStatsLine].numNew || varStats[i - firstVarStatsLine].numDel)
-            return false;
     return true;
 }
 
-// compare function to sort line offsets by line
-static int cmpLineOffsets(const void* off1, const void* off2)
+// compare function to sort symbols by line offsets of their creation
+static int cmpLifeTime(const void* p1, const void* p2)
 {
-    const LineOffset* loff1 = (const LineOffset*)off1;
-    const LineOffset* loff2 = (const LineOffset*)off2;
+    const LifeTime* lt1 = (const LifeTime*)p1;
+    const LifeTime* lt2 = (const LifeTime*)p2;
 
-    if (loff1->linnum == loff2->linnum)
-        return loff1->offset - loff2->offset;
-    return loff1->linnum - loff2->linnum;
+    return lt1->offCreate - lt2->offCreate;
+}
+
+// a parent scope contains the creation offset of the child scope
+static SYMIDX isParentScope(LifeTime* lifetimes, SYMIDX parent, SYMIDX si)
+{
+    if(parent < 0) // full function
+        return true;
+    return lifetimes[parent].offCreate <= lifetimes[si].offCreate &&
+           lifetimes[parent].offDestroy > lifetimes[si].offCreate;
+}
+
+// find a symbol that includes the creation of the given symbol as part of its life time
+static SYMIDX findParentScope(LifeTime* lifetimes, SYMIDX si)
+{
+    for(SYMIDX sj = si - 1; sj >= 0; --sj)
+        if(isParentScope(lifetimes, sj, si))
+           return sj;
+    return -1;
+}
+
+static int strHash(const char* s)
+{
+    int hash = 0;
+    for (; *s; s++)
+        hash = hash * 11 + *s;
+    return hash;
+}
+
+bool VarStatistics::hashSymbolIdentifiers(symtab_t* symtab)
```

Comment:
> getHash, get, in, del, keys, values, rehash, i.e. the names in aa.c. In fact, why not use the same implementation, just a non-virtual version of it? At some point, then it could be turned into a template and all the different ones merged.

I've renamed `strHash` to `getHash`. I think the data structure is too different from an AA, so I also changed the comment about it being a hash table.